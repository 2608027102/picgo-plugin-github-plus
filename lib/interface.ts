export type ImgType = {
  fileName: string;
  extname: string;
  imgUrl: string;
  width: number;
  height: number;
  type: string;
  id: string;
}

export type ImgZipType = {
  f: string,
  w: number,
  h: number,
  id: string
}
