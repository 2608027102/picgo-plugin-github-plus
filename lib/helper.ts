import dayjs from 'dayjs'
import { ImgType, ImgZipType } from './interface'
import { join } from 'path'
export function getNow () {
  return dayjs().format('YYYY-MM-DD hh:mm:ss')
}

export function zip (img: ImgType): ImgZipType {
  return {
    f: img.fileName,
    w: img.width,
    h: img.height,
    id: img.id
  }
}

export function unzip (
  img: ImgZipType,
  { type, domain, owner, repo, branch, path }
): ImgType {
  const { f: fileName, w: width, h: height, id } = img
  const extname = fileName.split('.').slice(-1)[0]
  const str = domain || 'https://raw.githubusercontent.com'
  const imgUrl = join(str, owner, repo, branch, path, fileName)
  return {
    fileName,
    width,
    height,
    id,
    extname,
    imgUrl,
    type
  }
}
