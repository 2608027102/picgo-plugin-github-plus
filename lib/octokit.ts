import Octokit from '@octokit/rest'
import { getNow } from './helper'
import { ImgType, PluginConfig } from './interface'
import { join } from 'path'

// export default octokit

class Octo {
  owner: string = ''
  repo: string = ''
  branch: string = ''
  path: string = ''
  token: string = ''
  customUrl: string = ''
  octokit = new Octokit()
  constructor ({ repo, branch, path = '', token, customUrl = '' }: PluginConfig) {
    const [ owner, r ] = repo.split('/')
    if (!r) throw new Error('Error in repo name')
    this.owner = owner
    this.repo = r
    this.branch = branch || 'master'
    this.path = path
    this.token = token
    this.customUrl = customUrl
  }
  authenticate () {
    this.octokit.authenticate({
      type: 'token',
      token: this.token
    })
  }

  updateDataJson ({ data, sha }) {
    const { owner, repo, branch } = this
    return this.octokit.repos.updateFile({
      owner,
      branch,
      repo,
      path: 'data.json',
      sha,
      message: `Sync dataJson by PicGo at ${getNow()}`,
      content: Buffer.from(JSON.stringify(data)).toString('base64')
    })
  }
  createDataJson (data) {
    const { owner, repo, branch } = this
    return this.octokit.repos.createFile({
      owner,
      repo,
      branch,
      path: 'data.json',
      message: `Sync dataJson by PicGo at ${getNow()}`,
      content: Buffer.from(JSON.stringify(data)).toString('base64')
    })
  }

  async getDataJson (): Promise<{
    lastSync: string;
    data: any[];
    sha?: string;
  }> {
    const { owner, repo, branch } = this
    const defaultRet = {
      lastSync: '',
      data: []
    }
    const d = await this.octokit.git.getTree({
      owner,
      repo,
      tree_sha: branch
    })
    const dataJson = d.data.tree.filter(each => each.path === 'data.json')[0]
    if (dataJson) {
      let content = await this.octokit.git.getBlob({
        owner,
        repo,
        file_sha: dataJson.sha
      })
      const buf = Buffer.from(content.data.content, content.data.encoding)
      const json = JSON.parse(buf.toString())
      return {
        ...defaultRet,
        ...json,
        sha: dataJson.sha
      }
    }
    return defaultRet
  }

  async upload (img) {
    const { owner, repo, branch, path = '' } = this
    const { fileName } = img
    const d = await this.octokit.repos.createFile({
      owner,
      repo,
      path: join(path, fileName),
      message: `Upload ${fileName} by picGo - ${getNow()}`,
      content: img.base64Image || Buffer.from(img.buffer).toString('base64'),
      branch
    })
    if (d) {
      return this.parseUrl(fileName)
    }
    throw d
  }
  parseUrl (fileName) {
    const { repo, path, customUrl } = this
    const domain = customUrl || 'https://raw.githubusercontent.com'
    return join(domain, repo, path, fileName)
  }
}

let ins: Octo = null

export function getIns (config: PluginConfig): Octo {
  if (ins) return ins
  return new Octo(config)
}
