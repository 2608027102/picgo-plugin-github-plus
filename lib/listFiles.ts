import { authenticate, getDataJson, upload } from './octokit'

const defaultConfig = {
  owner: 'zWingz',
  repo: 'imgur',
  branch: 'master'
}

authenticate('3ccd268c3d4679086c84baea7ad2e4c216e8f452')

getDataJson({
  owner: 'zWingz',
  repo: 'imgur',
  branch: 'master'
}).then(e => {
  console.log(e)
})

// upload({
//   ...defaultConfig,
//   path: 'data3.json',
//   message: 'Upload data.json',
//   content: Buffer.from(JSON.stringify([])).toString('base64')
// })
