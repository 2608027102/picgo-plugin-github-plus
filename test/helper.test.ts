import { pathJoin, zip, unzip, getNow } from '../lib/helper'
describe('test helper', () => {
  it('test pathJson', () => {
    expect(pathJoin('path', 'to', 'dir')).toEqual('path/to/dir')
    expect(pathJoin('path', '../to', 'dir')).toEqual('to/dir')
    expect(pathJoin('path', '啦啦啦', '中文')).toEqual('path/啦啦啦/中文')
  })

  it('test zip', () => {
    const str = zip({
      fileName: 'filename',
      width: 1000,
      height: 1000,
      id: 'fdsafdsafdsafasdf',
      sha: 'fooiuiouiouio'
    } as any)
    expect(str).toMatchSnapshot('zip')
  })
  it('test unzip', () => {
    const str = unzip({
      f: 'filename',
      w: 1000,
      h: 1000,
      id: 'fdsafdsafdsafasdf',
      s: 'fooiuiouiouio'
    } as any)
    expect(str).toMatchSnapshot('unzip')
  })
})
describe('test getNow', () => {
  const date = global.Date
  beforeEach(() => {
    const mockedDate = new Date(Date.UTC(2019, 1, 21, 16, 10, 20))
    global.Date = class extends Date {
      constructor () {
        super()
        return mockedDate
      }
    } as any
  })
  afterEach(() => {
    global.Date = date
  })
  it('test getNow', () => {
    const now = getNow()
    expect(now).toMatchSnapshot('now')
  })
})
