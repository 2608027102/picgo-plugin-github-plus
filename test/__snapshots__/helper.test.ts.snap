// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test getNow test getNow: now 1`] = `"2019-02-22 12:10:20"`;

exports[`test helper test unzip: unzip 1`] = `
Object {
  "extname": "filename",
  "fileName": "filename",
  "id": "fooiuiouiouio",
  "imgUrl": "",
  "sha": "fooiuiouiouio",
  "type": "",
}
`;

exports[`test helper test zip: zip 1`] = `
Object {
  "f": "filename",
  "s": "fooiuiouiouio",
}
`;
